src/block_url.rs

```diff
@@ -0,0 +1,140 @@
+use std::{error::Error, io};
+use url::Url;
+
+pub struct BlockedList {
+    root_id: i64,
+    sub_url: Vec<String>,
+    modified: String
+}
+
+pub async fn check_page_is_valid(client: &Client, connection: &Connection, url: &str) -> Result<bool, Box<dyn Error>> {
+    // TODO: check non HTML pages
+    if let Err(e) = check_blocked_list(connection, url).await { return Err(e) }
+    check_robots(client, url).await
+    // Ok(true)
+}
+
+async fn check_robots(client: &Client, url: &str) -> Result<bool, Box<dyn Error>> {
+    match parse_robots(client, url).await {
+        Ok(robots_parsed) => {
+            for line in robots_parsed {
+                let mut success = true;
+                if let (Some(k), Some(v)) = line {
+                    match k.trim() {
+                        "User-agent" => {
+                            // TODO: implement user-agent checking
+                        },
+                        "Disallow" => {
+                            let parsed_url = Url::parse(url)?;
+                            let url_subdir = parsed_url.path();
+        
+                            if url_subdir.trim_start_matches('/') == v.trim_start_matches('/') {
+                                success = false;
+                            }
+                            // TODO: check if url is within db, and if not, add it to database
+                        },
+                        _ => {}
+                    }
+                } else {
+                    info!("Robots.txt couldn't be parsed");
+                    success = false; // failsafe for broken robots.txt
+                }
+                if !success { return Ok(false); }
+            }
+            Ok(true)
+        },
+        Err(e) => Err(e)
+    }
+}
+
+async fn check_blocked_list(connection: &Connection, url: &str) -> Result<bool, Box<dyn Error>> {
+    // TODO: check timestamp and compare
+    let blocked_urls = get_blocked_list(connection, url).await?; 
+    let sub_path = get_subpath(url)?;
+
+    for url in blocked_urls.sub_url {
+        if url == sub_path { return Ok(false) }
+    }
+    Ok(true)
+}
+
+async fn get_blocked_list(connection: &Connection, url: &str) -> Result<BlockedList, Box<dyn Error>> {
+    let root_url = get_url_path(url, None)?;
+    let root: (i64, String) = connection.query_row(
+        "SELECT Id, Modified FROM RootUrl WHERE Url = ?1",
+        [&root_url],
+        |row| {
+            let id = row.get(0)?;
+            let modified = row.get(1)?;
+            Ok((id, modified))
+        },
+    )?;
+    
+    let blocked_list: Vec<String> = connection
+        .prepare("SELECT SubUrl FROM BlockedUrl WHERE RootId = ?")?
+        .query_map([root.0], |row| row.get::<_, String>(0))?
+        .collect::<Result<Vec<_>, _>>()?;
+    
+    Ok(BlockedList {
+        root_id: root.0,
+        sub_url: blocked_list,
+        modified: root.1,
+    })
+}
+
+fn get_subpath(url: &str) -> Result<String, Box<dyn std::error::Error>> {
```

Comment:
A "subpath" isn't really a standard name. At least, I'm not really familiar with the term, and had to look at the implementation to infer its meaning. I would consider annotating with some comments or docstrings, or consider a more standard name. I'm not an expert on this, but here's my suggestion. Feel free to correct me if I seem wrong.

```suggestion
/// Extracts the part of the url that defines the resource relative to the domain. The leading slash is omitted.
/// Example: `https://example.com/subdir/page` -> `subdir/page`
fn get_resource_path(url: &str) -> Result<String, Box<dyn std::error::Error>> {
```